src/styles/withStyles.d.ts

```diff
@@ -28,19 +30,22 @@ export interface WithStylesOptions {
 
 export type ClassNameMap<ClassKey extends string = string> = Record<ClassKey, string>;
 
-export interface WithStyles<ClassKey extends string = string> {
+export interface WithStyles<ClassKey extends string = string> extends Partial<WithTheme> {
   classes: ClassNameMap<ClassKey>;
-  theme?: Theme;
 }
 
 export interface StyledComponentProps<ClassKey extends string = string> {
   classes?: Partial<ClassNameMap<ClassKey>>;
   innerRef?: React.Ref<any>;
 }
 
+/*
+  The Omit<P, 'classes'> below (rather than the canonical Omit<P, WithStyles<ClassKey>>)
+  is used to allow interoperability with withTheme, which may provide the 'theme' prop.
+ */
 export default function withStyles<ClassKey extends string>(
   style: StyleRules<ClassKey> | StyleRulesCallback<ClassKey>,
   options?: WithStylesOptions,
-): <P>(
-  component: React.ComponentType<P & WithStyles<ClassKey>>,
-) => React.ComponentType<P & StyledComponentProps<ClassKey>>;
+): <P extends WithStyles<ClassKey>>(
+  component: React.ComponentType<P>,
+) => React.ComponentType<Omit<P, 'classes'> & StyledComponentProps<ClassKey>>;
```

Comment:
I think there's some merit here, but `P` should not be _required_ to included a `classes` prop. Instead it should be required that if `P` _does_ include a `classes` prop, it should have the right type. Something like this:

```ts
interface MaybeWithStyles extends Partial<WithStyles<ClassKey>> {
  [k: string]: any;
}

export default function withStyles<ClassKey extends string>(
  style: StyleRules<ClassKey> | StyleRulesCallback<ClassKey>,
  options?: WithStylesOptions,
): <P extends MaybeWithStyles<ClassKey>>(
  component: React.ComponentType<P & WithStyles<ClassKey>>,
) => React.ComponentType<Omit<P, 'classes'> & StyledComponentProps<ClassKey>>;
```